### PR TITLE
fix audt issues

### PR DESCRIPTION
**Pull request by danoctavian**

### CODE REVIEW

File: src/BaseVault.sol [hunk 1]
```diff
@@ -154,7 +154,7 @@ abstract contract BaseVault is IVault, ERC20PermitUpgradeable, AccessControlUpgr
         }
 
         uint256 ownerShares = balanceOf(owner);
-        uint256 maxAssets = convertToAssets(ownerShares);
+        uint256 maxAssets = previewRedeem(ownerShares);
```
Comment (danoctavian) [inline src/BaseVault.sol:157]: this way maxWithdraw accounts for fees.
From ERC4626:
„MUST return the maximum amount of assets that could be transferred from owner through withdraw and not cause a revert, which MUST NOT be higher than the actual maximum that would be accepted (it should underestimate if necessary).”

File: src/BaseVault.sol [hunk 2]
```diff
@@ -420,9 +421,11 @@ abstract contract BaseVault is IVault, ERC20PermitUpgradeable, AccessControlUpgr
             _spendAllowance(owner, caller, shares);
         }
 
+        // NOTE: burn shares before withdrawing the assets
+        _burn(owner, shares);
```
Comment (danoctavian) [inline src/BaseVault.sol:425]: implement the Checks-Effects-Interactions pattern.

File: src/BaseVault.sol [hunk 3]
```diff
@@ -674,7 +677,11 @@ abstract contract BaseVault is IVault, ERC20PermitUpgradeable, AccessControlUpgr
      * @dev This function iterates through the list of assets, gets their balances and rates,
      *      and updates the total assets denominated in the base asset.
      */
-    function processAccounting() public virtual {
+    function processAccounting() public virtual nonReentrant {
```
Comment (danoctavian) [inline src/BaseVault.sol:680]: add nonReentrant as well so an evil token cannot reenter through processAccounting while _withdraw is happening

**Review by dan13ram — APPROVED** [review]

File: src/BaseVault.sol (hunk 1 above)
Comment (dan13ram) [reply src/BaseVault.sol:157]: Got it. `KernelStrategy` currently doesn't satisfy this condition, and thats why we use the viewer instead to get maxWithdraw :)

Comment (danoctavian) [reply src/BaseVault.sol:157]: yes, we're ok with that for now there

Comment (danoctavian) [reply src/BaseVault.sol:157]: yes for now

**Pull request merged**